File: outdooer-frontend/src/api/permissions.js
// src/api/permissions.js
import api from './index';

/**
 * API module for permissions-related endpoints
 */
export const permissionsApi = {
  /**
   * Check if the current user has permission to perform an operation
   * 
   * @param {Object} data - Permission check data
   * @param {string} data.operation - The operation to check (e.g., 'create_expedition')
   * @param {number|null} data.resource_id - ID of the resource (optional)
   * @param {number|null} data.team_id - ID of the team (optional)
   * @returns {Promise<Object>} - Response with has_permission boolean
   */
  checkPermission: async (data) => {
    try {
      const response = await api.post('/permissions/check', data);
      return response.data;
    } catch (err) {
      console.error('Permission check error:', err);
      return { has_permission: false };
    }
  },
  
  /**
   * Get all permissions available to the current user
   * 
   * @returns {Promise<Object>} - Response with permissions data
   */
  getUserPermissions: async () => {
    try {
      const response = await api.get('/permissions/user');
      return response.data;
    } catch (err) {
      console.error('Error fetching user permissions:', err);
      return { permissions: [] };
    }
  },
  
  /**
   * Get all role configurations (global default permissions)
   * 
   * @returns {Promise<Object>} - Response with role configurations
   */
  getRoleConfigurations: async () => {
    try {
      const response = await api.get('/permissions/role-configurations');
      return response.data;
    } catch (err) {
      console.error('Error fetching role configurations:', err);
      return { role_configurations: {} };
    }
  },
  
  /**
   * Get team-specific permissions
   * 
   * @param {number} teamId - ID of the team
   * @returns {Promise<Object>} - Response with team permissions
   */
  getTeamPermissions: async (teamId) => {
    try {
      const response = await api.get(`/permissions/team/${teamId}/permissions`);
      return response.data;
    } catch (err) {
      console.error('Error fetching team permissions:', err);
      return { permissions: {} };
    }
  },
  
  /**
   * Update team-specific permissions
   * 
   * @param {number} teamId - ID of the team
   * @param {Object} permissions - Permissions object to update
   * @returns {Promise<Object>} - Response data
   */
  updateTeamPermissions: async (teamId, permissions) => {
    try {
      const response = await api.post(`/permissions/team/${teamId}/permissions`, { permissions });
      return response.data;
    } catch (err) {
      console.error('Error updating team permissions:', err);
      throw err;
    }
  },
  
  /**
   * Sync team permissions with default role configurations
   * 
   * @param {Object} data - Sync request data
   * @param {number} data.team_id - ID of the team
   * @returns {Promise<Object>} - Response data
   */
  syncTeamPermissions: async (data) => {
    try {
      const response = await api.post('/permissions/sync-permissions', data);
      return response.data;
    } catch (err) {
      console.error('Error syncing team permissions:', err);
      throw err;
    }
  }
};

export default permissionsApi;

File: outdooer-backend/app/middleware/permissions.py
# app/middleware/permissions.py
from functools import wraps
from flask import request, jsonify, g
from flask_jwt_extended import get_jwt_identity, verify_jwt_in_request
from app.models.team_member import TeamMember
from app.models.team_role_configuration import TeamRoleConfiguration
from app.models.team_role_permissions import TeamRolePermissions
from app.models.expedition import Expedition
from app.models.activity import Activity

def check_role_permission(operation):
    """
    Middleware decorator that checks if a user has permission to perform
    a specific operation based on their role level in the team.
    
    Args:
        operation (str): The operation to check permission for (e.g., 'create_expedition')
    """
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            # Verify JWT
            verify_jwt_in_request()
            current_user_id = get_jwt_identity()
            
            # Get team_id from request data
            data = request.get_json() or {}
            
            # For creation operations, team_id is in the request body
            team_id = data.get('team_id')
            
            # For update/delete operations, get the resource first to find the team_id
            resource_id = kwargs.get('id') or kwargs.get('expedition_id') or kwargs.get('activity_id')
            if resource_id and not team_id:
                if 'expedition' in operation:
                    resource = Expedition.query.get_or_404(resource_id)
                    team_id = resource.team_id
                elif 'activity' in operation:
                    resource = Activity.query.get_or_404(resource_id)
                    team_id = resource.team_id
            
            if not team_id:
                return jsonify({'error': 'Team ID not provided or resource not found'}), 400
            
            # Get user's role level in this team
            team_membership = TeamMember.query.filter_by(
                user_id=current_user_id,
                team_id=team_id
            ).first()
            
            if not team_membership:
                return jsonify({'error': 'You are not a member of this team'}), 403
            
            role_level = team_membership.role_level
            
            # First check team-specific permissions
            permission_key = operation
            team_permission = TeamRolePermissions.query.filter_by(
                team_id=team_id,
                role_level=role_level,
                permission_key=permission_key
            ).first()
            
            # If team-specific permission exists, use it
            if team_permission:
                has_permission = team_permission.is_enabled
            else:
                # Fall back to global role configuration
                global_permission = TeamRoleConfiguration.query.filter_by(
                    role_level=role_level,
                    operation=operation
                ).first()
                
                has_permission = global_permission.is_permitted if global_permission else False
            
            # Special case for update operations - Technical Guides can update their own resources
            if not has_permission and operation in ['update_expedition', 'update_activity'] and role_level == 3:
                if 'expedition' in operation:
                    resource = Expedition.query.get_or_404(resource_id)
                else:
                    resource = Activity.query.get_or_404(resource_id)
                
                # Technical Guides can update their own creations or ones they lead
                if resource.created_by == current_user_id or getattr(resource, 'leader_id', None) == current_user_id:
                    has_permission = True
            
            if not has_permission:
                role_names = {1: 'Master Guide', 2: 'Tactical Guide', 3: 'Technical Guide', 4: 'Base Guide'}
                role_name = role_names.get(role_level, f'Level {role_level} Guide')
                
                return jsonify({
                    'error': f'{role_name}s do not have permission to {operation.replace("_", " ")}'
                }), 403
            
            # Store role_level in flask.g for the route function
            g.role_level = role_level
            g.team_id = team_id
            
            return func(*args, **kwargs)
        return wrapper
    return decorator